components/brave_vpn/browser/connection/wireguard/win/wireguard_utils.cc

```diff
@@ -0,0 +1,211 @@
+      dns_servers.empty()) {
+    return absl::nullopt;
+  }
+  std::string config = kWireguardConfigTemplate;
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{client_private_key}",
+                                     client_private_key);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{server_public_key}",
+                                     server_public_key);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{vpn_server_hostname}",
+                                     vpn_server_hostname);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{mapped_ipv4_address}",
+                                     mapped_ipv4_address);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{dns_servers}", dns_servers);
+  return config;
+}
+
+bool EnableBraveVpnWireguardServiceImpl(const std::string& config) {
+  base::win::AssertComInitialized();
+  HRESULT hr = S_OK;
+  Microsoft::WRL::ComPtr<IBraveVpnWireguardManager> service;
+  CoCreateInstance(brave_vpn::GetBraveVpnWireguardServiceClsid(), nullptr,
+                   CLSCTX_LOCAL_SERVER,
+                   brave_vpn::GetBraveVpnWireguardServiceIid(),
+                   IID_PPV_ARGS_Helper(&service));
+  if (FAILED(hr)) {
+    VLOG(1) << "Unable to create IBraveVpnWireguardManager instance";
+    return false;
+  }
+
+  hr = CoSetProxyBlanket(
+      service.Get(), RPC_C_AUTHN_DEFAULT, RPC_C_AUTHZ_DEFAULT,
+      COLE_DEFAULT_PRINCIPAL, RPC_C_AUTHN_LEVEL_PKT_PRIVACY,
+      RPC_C_IMP_LEVEL_IMPERSONATE, nullptr, EOAC_DYNAMIC_CLOAKING);
+  if (FAILED(hr)) {
+    VLOG(1) << "Unable to call EnableVpn interface";
+    return false;
+  }
+  std::string encoded_config;
+  base::Base64Encode(config, &encoded_config);
+  DWORD error_code = 0;
+  if (FAILED(service->EnableVpn(base::UTF8ToWide(encoded_config).c_str(),
+                                &error_code))) {
+    VLOG(1) << "Unable to call EnableVpn interface";
+    return false;
+  }
+  return error_code == 0;
+}
+
+void EnableBraveVpnWireguardService(const std::string& config,
+                                    wireguard::BooleanCallback callback) {
+  base::ThreadPool::CreateCOMSTATaskRunner(
+      {base::MayBlock(), base::WithBaseSyncPrimitives(),
+       base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN},
+      base::SingleThreadTaskRunnerThreadMode::DEDICATED)
+      ->PostTaskAndReplyWithResult(
+          FROM_HERE,
+          base::BindOnce(&EnableBraveVpnWireguardServiceImpl, config),
+          std::move(callback));
+}
+
+bool DisableBraveVpnWireguardServiceImpl() {
+  base::win::AssertComInitialized();
+  HRESULT hr = S_OK;
+  Microsoft::WRL::ComPtr<IBraveVpnWireguardManager> service;
+  CoCreateInstance(brave_vpn::GetBraveVpnWireguardServiceClsid(), nullptr,
+                   CLSCTX_LOCAL_SERVER,
+                   brave_vpn::GetBraveVpnWireguardServiceIid(),
+                   IID_PPV_ARGS_Helper(&service));
+  if (FAILED(hr)) {
```

Comment:
Same question: `hr = CoCreateInstance(...);`?

Reply:
And there's another one in the `WireguardGenerateKeypairImpl()` function below.

Reply:
done